### DISCUSSION TITLE
Error unsupported collating sort order 3082

### DISCUSSION THREAD

**Discussion by rbenrax** [jahlborn/jackcess#15] [General]
I have solved the problem of unsupported collating sort order 3082" error with the following code in the IndexData class, in method createSpecialEntry:

    switch(col.getType()) {
    case TEXT:
    case MEMO:
      ColumnImpl.SortOrder sortOrder = col.getTextSortOrder();
      
      if(ColumnImpl.GENERAL_SORT_ORDER.equals(sortOrder)) {
        return new GenTextColumnDescriptor(col, flags);
      }
      if(ColumnImpl.GENERAL_LEGACY_SORT_ORDER.equals(sortOrder)) {
        return new GenLegTextColumnDescriptor(col, flags);
      }
      if(ColumnImpl.GENERAL_97_SORT_ORDER.equals(sortOrder)) {
        return new Gen97TextColumnDescriptor(col, flags);
      }
      
      // Rafa (04/10/2024) Sortorder collation error
      if (sortOrder.getValue()==(short)3082 && sortOrder.getVersion()==(short)0){
        return new GenLegTextColumnDescriptor(col, flags);
      }

Last block of code.

It is possible that there is a more orthodox way of implementation, for now this is how it works and the error is avoided.

**Comment by jahlborn** [comment]
> I have solved the problem of unsupported collating sort order 3082" error with the following code in the IndexData class, in method createSpecialEntry:
> 
> ```
> switch(col.getType()) {
> case TEXT:
> case MEMO:
>   ColumnImpl.SortOrder sortOrder = col.getTextSortOrder();
>   
>   if(ColumnImpl.GENERAL_SORT_ORDER.equals(sortOrder)) {
>     return new GenTextColumnDescriptor(col, flags);
>   }
>   if(ColumnImpl.GENERAL_LEGACY_SORT_ORDER.equals(sortOrder)) {
>     return new GenLegTextColumnDescriptor(col, flags);
>   }
>   if(ColumnImpl.GENERAL_97_SORT_ORDER.equals(sortOrder)) {
>     return new Gen97TextColumnDescriptor(col, flags);
>   }
>   
>   // Rafa (04/10/2024) Sortorder collation error
>   if (sortOrder.getValue()==(short)3082 && sortOrder.getVersion()==(short)0){
>     return new GenLegTextColumnDescriptor(col, flags);
>   }
> ```
> 
> Last block of code.
> 
> It is possible that there is a more orthodox way of implementation, for now this is how it works and the error is avoided.

so that may stop the exception from happening, but it's probable that you are creating an access database which is going to have invalid data.  those codes correspond to the "sort order" of the database.  the sort order defines how to translate characters to code values which can be used to correctly order string values.  currently, only the "general" variants are supported.  "3802" corresponds to "Spanish - Spain (Modern Sort)".  Jackcess may write the database without error, but it's highly likely that when you open the database in access, it will complain that the data is invalid (especially if you try to modify the data).

**Comment by rbenrax** [comment]
I have carried out intensive tests with specific characters of the Spanish - Spain (Modern Sort) sorting, and the result has been correct, it has allowed me to modify the data from Access and from Java, the database is encoded in Unicode (UTF-16 ) just like Java, the code conversion should not be affected, although it is possible that the ordering of the Index is not completely correct.

But if there are no plans to implement other collation oreders, this is better than nothing to me.

**Comment by jahlborn** [comment]
> I have carried out intensive tests with specific characters of the Spanish - Spain (Modern Sort) sorting, and the result has been correct, it has allowed me to modify the data from Access and from Java, the database is encoded in Unicode (UTF-16 ) just like Java, the code conversion should not be affected, although it is possible that the ordering of the Index is not completely correct.
> 
> But if there are no plans to implement other collation oreders, this is better than nothing to me.

i don't currently have plans for other collation orders because it's a lot of effort to thoroughly test and verify each collation order (assuming that i have the means to do it).

the way that i typically test a collation order is to create a table with string entries which contain every character of the given character set, usually sandwiched between two other characters (e.g. for english "aaa", "aba", "aca"... "aXa").  i open the table in access, apply the collation order to the database, and create an index on the column with the data.  then i open the table in jackcess, grab each row and see if the index entry created by jackcess matches the index entry created in the table currently (from access).
